### ISSUE TITLE
[FEA] Function to "pack" a table into a single buffer

### ISSUE THREAD

**Issue by jrhemstad** [rapidsai/cudf#3793]
**Is your feature request related to a problem? Please describe.**

I would like to be able to create a copy of a `table_view` where the storage for the constituent column is "packed" into a single allocation. 

**Describe the solution you'd like**

An API something like this:
```
std::pair<rmm::device_buffer, table_view> pack(table_view const& t);
```

Where the returned `device_buffer` contains a copy of the contents of `t` in a single allocation and the returned `table_view` points into that buffer. 

**Additional context**

This functionality is effectively already developed for `contiguous_split` in the[ `alloc_and_copy` ](https://github.com/rapidsai/cudf/blob/930810ae3f26fe49eddee70f9aedc6ff10f3ba43/cpp/src/copying/contiguous_split.cu#L175) function. The ask is just to wrap and expose this as a public API.

**Comment by felipeblazing** [comment]
We would love to have this. Shouldn't we have somethin that is alse like
```
std::unique_ptr<table> unpack(const rmm::device_buffer & buffer, unsigned long long size); 
```

**Comment by jrhemstad** [comment]
> We would love to have this. Shouldn't we have somethin that is alse like
> 
> ```
> std::unique_ptr<table> unpack(const rmm::device_buffer & buffer, unsigned long long size); 
> ```

I think it'd be better to return a `table_view` that points into the returned `device_buffer`. I updated the issue.

**Comment by jakirkham** [comment]
This would also be useful from the UCX side as we could serialize a larger contiguous piece of memory and then unpack it later so that individual columns could be returned to RMM when no longer needed.

cc @quasiben

**Comment by jrhemstad** [comment]
@jakirkham there was some offline conversation with @jlowe @kkraus14 @nvdbaranec about how we'd serialize a "packed" table for communication. 

The gist of it is, when you "pack" a table, you get a `device_buffer` and a `table_view` that points into that `device_buffer`. If you were to simply communicate the `device_buffer` and `table_view` as-is, the pointers in the `table_view` on the receiving side wouldn't be correct because they'd refer to memory locations on the sending side.

It's pretty easy to update the pointers on the receiving side if you know what the base pointers were on the sending side. It's just a bunch of pointer offset arithmetic.  

We're trying to figure out a common set of serialize/deserialize primitives for libcudf that both Dask and Spark can use. Could you weigh in and help to define that the serialized layout should look like?

**Comment by quasiben** [comment]
@madsbk do you have opinions here as well on things ucx-py could be/should be doing ?

**Comment by madsbk** [comment]
> @madsbk do you have opinions here as well on things ucx-py could be/should be doing ?

I don't think that ucx-py should support any serialization directly. 
But it would be great if cuDF could serialize/deserialize multiple columns to/from a single `device_buffer`.

**Comment by jakirkham** [comment]
> > @madsbk do you have opinions here as well on things ucx-py could be/should be doing ?
> 
> I don't think that ucx-py should support any serialization directly.
> But it would be great if cuDF could serialize/deserialize multiple columns to/from a single `device_buffer`.

Yeah I was thinking we might do this in Distributed before it got to UCX-Py. More-or-less the same as PR ( https://github.com/dask/distributed/pull/3453 ).

**Comment by jakirkham** [comment]
FWIW I think from the Dask side w.r.t. cuDF we would be well served for most cases by having something we could [plug into cuDF `"cuda"` serialization]( https://github.com/rapidsai/cudf/blob/branch-0.14/python/cudf/cudf/comm/serialize.py#L25-L31 ). We could then reuse this for transmission with TCP or UCX and spilling since they all go through that code path now.

There are some other objects we would like similar functionality (like CuPy sparse arrays). Though I think other Numba and CuPy objects are already working with contiguous memory chunks to the extent possible.

**Comment by jrhemstad** [comment]
This was closed by #7096
